### PR TITLE
WhatsApp Cloud App

### PR DESCRIPTION
**Pull request by Sandro-Meireles**

### CODE REVIEW

File: marketplace/core/types/channels/whatsapp_base/mixins.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        except (ValueError, TypeError):
+            self._raise(key)
+
+    def _raise(self, field: str):
+        raise ValidationError(self.ERROR_MESSAGE.format(field))
+
+
+class WhatsAppConversationsMixin(object):
+    @action(detail=True, methods=["GET"], permission_classes=[ProjectViewPermission])
+    def conversations(self, request: "Request", **kwargs) -> Response:
```
Comment (pauloabreu) [inline marketplace/core/types/channels/whatsapp_base/mixins.py:57]: Looking at rest_framework viewsets, i think parameter request is a Request object, not a string

File: marketplace/core/types/channels/whatsapp/views.py [hunk 2]
```diff
@@ -13,153 +11,41 @@
...
-            return Response(serializer.data)
-
-        except FacebookApiException:
-            raise ValidationError(
-                "There was a problem requesting the On-Premise API, check if your authentication token is correct"
-            )
+        return dict(base_url=base_url, auth_token=auth_token)
 
     @action(detail=False, methods=["GET"], url_name="shared-wabas", url_path="shared-wabas")
     def shared_wabas(self, request: "Request", **kwargs):
```
[the remark this answers is not in the document]
Comment (pauloabreu) [reply marketplace/core/types/channels/whatsapp/views.py:51]: Has anothers lines with same questions, i'll not repeat, if this is a real problem, do a Ctrl + F

File: marketplace/core/types/channels/whatsapp_cloud/requests.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        return settings.WHATSAPP_API_URL + f"/{self._phone_number_id}" + self._endpoint
+
+    @property
+    def _headers(self) -> dict:
+        access_token = settings.WHATSAPP_SYSTEM_USER_ACCESS_TOKEN
+        return {"Content-Type": "application/json", "Authorization": f"Bearer {access_token}"}
+
+    def get_profile(self):
+        response = requests.get(self._url, params=self._fields, headers=self._headers)
+        content = response.json().get("data", [])[0]
```
Comment (pauloabreu) [inline marketplace/core/types/channels/whatsapp_cloud/requests.py:30]: I think you can change:
`content = response.json().get("data", [])[0]`
to:
`content = response.json().get("data", [{}])[0]`
Its prevents a IndexError

File: marketplace/core/types/channels/whatsapp_cloud/views.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+            raise ValidationError("The phone number is not configured")
+
+        return dict(phone_number_id=phone_numbrer_id)
+
+    def get_queryset(self):
+        return super().get_queryset().filter(code=self.type_class.code)
+
+    def destroy(self, request, *args, **kwargs) -> Response:
+        return Response(
+            "This channel cannot be deleted", status=status.HTTP_403_FORBIDDEN
```
[the remark this answers is not in the document]
Comment (pauloabreu) [reply marketplace/core/types/channels/whatsapp_cloud/views.py:53]: https://github.com/weni-ai/weni-integrations-engine/pull/148

**Comment by Sandro-Meireles** [comment]
Run black on the entire project to fix lint failures.

File: marketplace/core/types/channels/whatsapp_cloud/facades.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+    def __init__(self, phone_number_id: "str") -> None:
+        self._profile_api = CloudProfileRequest(phone_number_id)
+        self._photo_api = PhotoAPIRequest(phone_number_id)
+
+    def get_profile(self):
+        profile = self._profile_api.get_profile()
+        vertical = profile["business"]["vertical"]
+
+        if vertical is not None:
```
[the remark this answers is not in the document]
Comment (pauloabreu) [reply marketplace/core/types/channels/whatsapp_cloud/facades.py:36]: solving `content = response.json().get("data", [{}])[0]` inside of _profile_api.get_profile() exclude possibility of `profile["business"]["vertical"]` generete a KeyError

**Review by pauloabreu — APPROVED** [review]

**Pull request merged**